Prevent automatic cuda init in init_rpc

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #80180

Fixes #80141

Only initialize cuda if there are devices specified in `init_rpc`

Differential Revision: [D37458309](https://our.internmc.facebook.com/intern/diff/D37458309)

H-Huang committed:
Prevent automatic cuda init in init_rpc

[ghstack-poisoned]

H-Huang committed:
Update on "Prevent automatic cuda init in init_rpc"


Fixes #80141

Only initialize cuda if there are devices specified in `init_rpc` 

[ghstack-poisoned]

rohan-varma commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -262,6 +262,8 @@ def _get_device_infos():
     opts = agent._get_backend_options()
     device_count = torch.cuda.device_count()
+    if torch.cuda.is_available() and opts.devices:
+        torch.cuda.init()
```
Is there a reason we add torch.cuda.init() here in addition to the call in the tensorpipe backend handler? Could we possibly have it in just one place?

H-Huang replied:
One is for statically sized groups, the other is for dynamic/elastic groups (devices are updated as workers join, therefore we need to do another cuda init if new devices are added) so they have different initialization paths and we can't put it in the same place for now.

rohan-varma reviewed (commented):
LGTM, just one question for your consideration

rohan-varma reviewed (approved):
thanks for the fix!

H-Huang committed:
Update on "Prevent automatic cuda init in init_rpc"


Fixes #80141

Only initialize cuda if there are devices specified in `init_rpc`

Differential Revision: [D37458309](https://our.internmc.facebook.com/intern/diff/D37458309)

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37458309).

H-Huang commented:
@pytorchbot rebase

H-Huang committed:
Update on "Prevent automatic cuda init in init_rpc"


Fixes #80141

Only initialize cuda if there are devices specified in `init_rpc`

Differential Revision: [D37458309](https://our.internmc.facebook.com/intern/diff/D37458309)

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37458309).

H-Huang commented:
@pytorchbot merge
